src/consensus/params.cpp

```diff
@@ -133,21 +133,15 @@ namespace Consensus {
     std::variant<FundingStream, FundingStreamError> FundingStream::ValidateFundingStream(
         const Consensus::Params& params,
         const int startHeight,
-        const int endHeight,
+        const int nFundingPeriodLength,
         const std::vector<FundingStreamAddress>& addresses
     ) {
+        // Funding stream periods may not start before Canopy
         if (!params.NetworkUpgradeActive(startHeight, Consensus::UPGRADE_CANOPY)) {
             return FundingStreamError::CANOPY_NOT_ACTIVE;
         }
 
-        if (endHeight < startHeight) {
-            return FundingStreamError::ILLEGAL_RANGE;
-        }
-
-        if (params.FundingPeriodIndex(startHeight, endHeight - 1) >= addresses.size()) {
-            return FundingStreamError::INSUFFICIENT_ADDRESSES;
-        }
-
+        auto endHeight = startHeight + (nFundingPeriodLength * addresses.size());
```

Comment:
This doesn't align with ZIP 207. ZIP 207 allows for any start or end height, and requires that there are sufficient addresses to cover that interval. This change restricts the allowed values of `endHeight` while still allowing `startHeight` to be arbitrary (`endHeight` must always be the same distance from an `AddressChangeInterval` boundary as `startHeight` is), and instead specifies the funding streams in terms of an exact number of `AddressChangeInterval`s. 

This change happens to be fine, because all existing defined funding streams are exact multiples of `AddressChangeInterval` (and I think they have `startHeight`s that also happen to align). We also did not observe anyone making use of the "direct grant" option of ZIP 1014 (that would have resulted in an existing FS being split into two ranges at some potentially-arbitrary height). So we could decide to retroactively declare FSs as being limited in this way. However, that would necessitate updates to both ZIP 207 and ZIP 214 to redefine the existing FSs appropriately (and the associated confirmation that we didn't break anything or introduce an off-by-one error).

In the interest of "make as few changes to `zcashd` as possible", let's keep things the way they were, and do the "number of lockbox recipients" check at runtime.